content/en/blog/_posts/2023-12-05-image-filesystem.md

```diff
@@ -0,0 +1,170 @@
+
+CRI-O uses a storage configuration file to control how the container runtime stores persistent and temporary data.
+CRI-O utilizes the [storage library](https://github.com/containers/storage).  
+Some Linux distributions have a manual entry for storage (`man 5 containers-storage.conf`).
+The main configuration for storage is located in `/etc/containers/storage.conf` and one can control the location for temporary data and the root directory.  
+The root directory is where CRI-O stores the persistent data.
+
+```toml
+[storage]
+# Default storage driver
+driver = "overlay"
+# Temporary storage location
+runroot = "/var/run/containers/storage"
+# Primary read/write location of container storage 
+graphroot = "/var/lib/containers/storage"
+```
+
+- graphroot
+  - Persistent data stored from the container runtime
+  - If SELinux is enabled, this must match the `/var/lib/containers/storage`
+- runroot
+  - Temporary read/write access for container
+  - Recommended to have this on a temporary filesystem
+
+A quick way to relabel your graphroot directory to match `/var/lib/containers/storage`.
+
+```bash
+semanage fcontext -a -e /var/lib/containers/storage /YOUR_STORAGE_PATH
+restorecon -R -v /YOUR_STORAGE_PATH
+```
+
+### containerd
+
+containerd uses a toml configuration file to control where persistent and ephemeral data is stored.
+The default path for the config file is located at `/etc/containerd/config.toml`.
+
+The relevant fields for containerd storage are `root` and `state`.
+
+- root
+  - The root directory for containerd metadata
+  - Default is `/var/lib/containerd`
+  - Root also requires selinux labels if your OS requires it
+- state
+  - Temporary data for containerd
+  - Default is `/run/containerd`
+
+## Kubernetes Node Pressure Eviction
+
+Kubernetes will automatically detect if the container filesystem is split from the node filesystem. When one separates the filesystem, Kubernetes is responsible for monitoring both the node filesystem and the container runtime filesystem.
+Kubernetes documentation refers to the node filesystem and the container runtime filesystem as nodefs and imagefs.
+If either nodefs or the imagefs are running out of disk space, then the overall node is considered to have disk pressure.
+Kubernetes will first reclaim space by deleting unusued containers and images, and then it will resort to evicting pods.
+If one has a imagefs, then Kubernetes will garbage collect unused images on imagefs and will remove dead containers and pods from the nodefs.
+If there is only a nodefs, then Kubernetes garbage collection includes dead containers, dead pods and unused images.
+
+Kubernetes allows more configurations for determining if your disk is full.  
+The eviction manager in Kubelet defines API fields to signal pressure.
+For filesystems, the relevant API are `nodefs.available`, `nodefs.inodesfree`, `imagefs.available`, and `imagefs.inodesfree`.
+If there is no container runtime filesystem then imagefs is ignored.
```

Comment:
(nit)
```suggestion
If there is no container image filesystem then `imagefs.*` is ignored.
```
?

Reply:
I think my suggestion is correct for the blog post. I want to call this the container runtime filesystem. In this blog I don't talk about separating the images and containers but I will be working on next so want to not call this image filesystem at the moment. Esp because both images/containers are stored on imagefs.

Reply:
OK, then I recommend you clarify this a lot more. Readers may infer (like I did) that you meant “container image filesystem”.

I _think_ this stems from the kubelet being written to assume one block-backed filesystem for the node and no mount points other than system filesystems (`/proc`, `/sys` etc). However, that assumption has been gradually unpeeled since Kubernetes 1.0 without addressing the terminology and its implications. It has ended up hard to teach.

My advice: introduce the concepts (_container image filesystem_, _container runtime filesystem_, etc) explicitly before you reference them, and define them enough that it's clear.

Reply:
I think we should just reference container runtime filesystem as KEP-4191 is only alpha and I'd like to introduce that concept in a follow up post. 

I tried to add a sentence about the POSIX difference between root and k8s root.

Reply:
I stand by the advice in https://github.com/kubernetes/website/pull/43971#discussion_r1438880825

Reply:
I updated this to be a bit more clear.